design/crypto_primitives.rs

```diff
@@ -0,0 +1,171 @@
+            from()
+        }
+        /// Occurs when we can't decrypt a message or verify the signature.
+        DecryptVerify(_e: ()) {
+            display("error decrypting/verifying message")
+            from()
+        }
+        /// Occurs in case of an error during initialisation.
+        InitError(e: i32) {
+            display("error while initialising random generator")
+            from()
+        }
+        /// Occurs when we fail to derive encryption key from password.
+        DeriveKey {
+            display("error deriving encryption key from password")
+        }
+    }
+}
+
+/// Derives encryption key from plaintext password.
+pub fn key_from_password(password: &[u8], salt: &[u8]) -> Result<SymmetricKey, Error>;
+
+/// Use `encrypt_key_pair()` to generate public and private keys pair.
+// Make it `Copy`able?
+#[derive(Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize, Clone)]
+pub struct PublicEncryptKey {
+    encrypt: box_::PublicKey,
+}
+
+impl PublicEncryptKey {
+    /// Construct public key from bytes. Useful when it was serialized before.
+    pub fn from_bytes(public_key: [u8; PUBLIC_ENCRYPT_KEY_BYTES]) -> Self
+
+    /// For anyone who wants to store public key.
+    pub fn as_bytes(self) -> [u8; PUBLIC_ENCRYPT_KEY_BYTES]
+
+    pub fn encrypt_anonymous<T: Serialize>(&self, plaintext: &T) -> Result<Vec<u8>, Error>
+
+    pub fn encrypt_anonymous_bytes(&self, plaintext: &[u8]) -> Vec<u8> {
+}
+
+/// Store secret key in `Arc` to discourage cloning it. Makes it more secure by not duplicating
+/// encryption keys in memory.
+/// Use `encrypt_key_pair()` to generate public and private keys pair.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct SecretEncryptKey {
+    inner: Arc<SecretEncryptKeyInner>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+struct SecretEncryptKeyInner {
+    encrypt: box_::SecretKey,
+}
+
+impl SecretEncryptKey {
+    /// Construct secret key from given bytes. Useful when secret key was serialized before.
+    pub fn from_bytes(secret_key: [u8; SECRET_ENCRYPT_KEY_BYTES]) -> Self
+
+    /// Derive shared key for symmetric data encryption
+    pub fn shared_secret(&self, their_pk: &PublicEncryptKey) -> SharedSecretKey
+
+    /// For anyone who wants to store secret key.
+    pub fn as_bytes(self) -> [u8; SECRET_ENCRYPT_KEY_BYTES]
+
+    pub fn decrypt_anonymous<T: Serialize + DeserializeOwned>(&self, ciphertext: &[u8], my_pk: &PublicEncryptKey) -> Result<T, Error>
+
+    pub fn decrypt_anonymous_bytes(&self, ciphertext: &[u8], my_pk: &PublicEncryptKey) -> Result<Vec<u8>, Error>
+}
+
+/// Randomly generates a secret key and a corresponding public key.<Paste>
+pub fn encrypt_key_pair() -> (PublicEncryptKey, SecretEncryptKey)
+
+#[derive(Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize, Clone)]
+pub struct PublicSignKey {
+    sign: sign::PublicKey,
+}
+
+impl PublicSignKey {
+    pub fn verify_detached(&self, signature: &Signature, data: &[u8]) -> bool {
```

Comment:
Should there be a constsructor that takes a `PublicKey`?

Reply:
there's `from_bytes()` below. Or did you mean something else ?